src/LAMMPS.jl

```diff
@@ -361,19 +361,149 @@ function extract_variable(lmp::LMP, name::String, group=nothing)
     end
 end
 
-function gather_atoms(lmp::LMP, name, T, count)
-    if T === Int32
-        dtype = 0
-    elseif T === Float64
-        dtype = 1
+@deprecate gather_atoms(lmp::LMP, name, T, count) gather(lmp, name, T)
+
+
+"""
+    gather(lmp::LMP, name::String, T::Union{Type{Int32}, Type{Float64}}, ids::Union{Nothing, Array{Int32}}=nothing)
+
+Gather the named per-atom, per-atom fix, per-atom compute, or fix property/atom-based entities from all processes.
+By default (when `ids=nothing`), this method collects data from all atoms in consecutive order according to their IDs.
+The optional parameter `ids` determines for which subset of atoms the requested data will be gathered. The returned data will then be ordered according to `ids`
+
+Compute entities have the prefix `c_`, fix entities use the prefix `f_`, and per-atom entites have no prefix.
+
+The returned Array is decoupled from the internal state of the LAMMPS instance.
+
+!!! warning "Type Verification"
+    Due to how the underlying C-API works, it's not possible to verify the element data-type of fix or compute style data.
+    Supplying the wrong data-type will not throw an error but will result in nonsensical output
+
+!!! warning "ids"
+    The optional parameter `ids` only works, if there is a map defined. For example by doing:
+    `command(lmp, "atom_modify map yes")`
+    However, LAMMPS only issues a warning if that's the case, which unfortuately cannot be detected through the underlying API.
+    Starting form LAMMPS version `17 Apr 2024` this should no longer be an issue, as LAMMPS then throws an error instead of a warning.
+"""
+function gather(lmp::LMP, name::String, T::Union{Type{Int32}, Type{Float64}}, ids::Union{Nothing, Array{Int32}}=nothing)
+    @assert name != "mass" "scattering/gathering mass is currently not supported! Use `extract_atom()` instead."
```

Comment:
```suggestion
    @assert name != "mass" "scattering/gathering mass is currently not supported! Use `gather_atoms()` instead."
```

Except that we deprecated that function?

Instead of `@assert`, `error` is more appropriate. `@asserts` may be removed by the compiler (in future versions of Julia).

Reply:
Oh `extract` does still exist. What is the difference between extract and gather?

Reply:
Extract gives you the pointer to the memory inside the Lammps instance. `extract_atom()` takes that pointer and wrapps that in an array and then returns that. So when you change values in the extracted array, it changes the values inside the lammps insance. Also, if you have multiple MPI ranks, you only get the data that's on that one MPI rank, at least according to the documentation. For now, I haven't run Lammps with more than one MPI rank though.

gather crashes LAMMPS and Julia when trying to gather masses - probably because that's not a proper per atom entity - so I've decided to dissalow using it with gather. I might take a closer look into that in the future so that we can reenable it if there is a way to gather masses that doesn't result in a crash.

Maybe I'll implement a fallback to `extract_atom()` that copies the data before returning it

Reply:
Yeah that fallback is #37

In https://docs.lammps.org/Classes_atom.html#_CPPv4N9LAMMPS_NS4Atom7extractEPKc 
it notes: `per-type mass. This array is NOT a per-atom array but of length ntypes+1, element 0 is ignored.`

But doesn't need to be fixed here.

Reply:
Should I change all assertions to errors?